### PR TITLE
Test/multisig wallet poc

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

We need to validate our understanding of how Native Scripts work & whether/how we can apply them to Shared Wallet work.

# Proposed Solution

Create a POC using the cardano-js-sdk of a multi signature wallet and show its usage in a e2e test.

This POC introduces two classes for demonstration:

**MultiSigWallet:** This kind of wallet can be created with a set of public keys, the wallet internally creates a script using said keys. The keys are sorted so the script can be recreated the same way every time, so every participant only needs to known which keys are involved in the multsignature wallet and they can recreate the wallet individually. The wallet then uses the script to create a credential, this credential is used for both the payment part and the staking part of the wallet address. This means users can control the funds and the stake rewards with the same script.

The wallet only supports two type of transactions. transferFunds and delegate, as these actions were enough to demonstrate the concept, but transaction of any nature can be included. The wallet also allows a particular user to sign a trnasaction and submit it to the network.

**MultiSigTx:** This class keeps track of the transaction and the participants involved in the multigisnature wallet. It also ahs some utility methods to determine which participates have signed and which signatures are missing. The transaction can also be encoded/decoded to CBOR for easy trasmition between participants.

This PR also includes 2 e2e tests.

**Simple fund transfer:** In this test 3 participants share a wallet Alice, Bob and Charlotte. Alice inanities the transaction and sings it. She then serializes the MultiSigTx and sends it to Bob (transmission method out of the scope of the POC), Bob receives the transaction, verify its contents, signs it, and finally checks whether any participant signature is missing (and which participant), he sees Charlotte signature is missing, so he serializes the transaction (now with his signature) and relays it to Charlotte. Charlotte receives the transaction, deserializes it and check the contents, she then signs it and sees that all required participants have already signed so she submits the transaction to the network. The last step of the test is to verify that the transaction made it on chain and the funds arrived back to the faucet wallet.

This test demonstrates the basic use case of a multi signature wallet.

**Delegate and claim rewards:** In this test, the three participants build and sign a transaction to delegate to one of the pools on our local network, after the delegation has been confirmed and some epochs have elapse, we use the faucet wallet to spam transaction on the network in order to generate some transaction fees that can be distribute as rewards to the delegators. We then check our reward account and wait until rewards are available, after we confirm the rewards are present, we send a transaction which consumes the rewards, the final step is to confirm that the transaction made it on chain, and that all rewards were successfully withdrawn from the account.

All transaction related to the multisig wallet used as a sole witness the script.

# Bug fixes

When developing the POC I encounter two problems, one was a bug that was preventing the correct creation of credentials using script hashes (fixed here https://github.com/input-output-hk/cardano-js-sdk/commit/e249448bfd31f1d1d24a0f215761b0b785e254ac).

The second problem is that our Certificate core types, for StakeRegistration, StakeDeregistration and Stake always assume we are using a stakeKeyHash, so they had to be refactored to be inclusive of all types of credentials. This ended being a big refactor.

### CODE REVIEW

File: packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        filter((value) => value > 0n),
+        take(1)
+      )
+    );
+
+    expect(multiSigWalletRewardBalance).toBeGreaterThan(0n);
+
+    logger.info(`Generated rewards: ${multiSigWalletRewardBalance} tLovelace`);
+
+    tx = await aliceMultiSigWallet.transferFunds(faucetAddress, { coins: 2_000_000n });
```
Comment (mkazlauskas) [inline packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts:284]: This is an important part of the test, best to be explicit about it
```suggestion
    tx = await aliceMultiSigWallet.transferFunds(faucetAddress, { coins: 2_000_000n });
    expect(tx.withdrawals.length).toBeGreaterThan(0);
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    );
+
+    const multiSigParticipants = [alicePubKey, bobPubKey, charlottePubKey];
+
+    const aliceMultiSigWallet = await createMultiSignWallet(aliceKeyAgent, faucetWallet, multiSigParticipants);
+    const bobMultiSigWallet = await createMultiSignWallet(bobKeyAgent, faucetWallet, multiSigParticipants);
+    const charlotteMultiSigWallet = await createMultiSignWallet(charlotteKeyAgent, faucetWallet, multiSigParticipants);
+
+    expect(aliceMultiSigWallet.getPaymentAddress()).toEqual(bobMultiSigWallet.getPaymentAddress());
+    expect(aliceMultiSigWallet.getPaymentAddress()).toEqual(charlotteMultiSigWallet.getPaymentAddress());
```
[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/992/commits/8664427d551aebded9b0979446284bf483dc3b91

File: packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    const { keyAgent: charlotteKeyAgent, pubKey: charlottePubKey } = await getKeyAgent(
+      charlotteMnemonics,
+      faucetWallet
+    );
+
+    const multiSigParticipants = [alicePubKey, bobPubKey, charlottePubKey];
+
+    const aliceMultiSigWallet = await createMultiSignWallet(aliceKeyAgent, faucetWallet, multiSigParticipants);
+    const bobMultiSigWallet = await createMultiSignWallet(bobKeyAgent, faucetWallet, multiSigParticipants);
+    const charlotteMultiSigWallet = await createMultiSignWallet(charlotteKeyAgent, faucetWallet, multiSigParticipants);
```
[the remark this answers is not in the document]
Comment (AngelCastilloB) [reply packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/992/commits/8664427d551aebded9b0979446284bf483dc3b91

File: packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts (hunk 1 above)
Comment (AngelCastilloB) [reply packages/e2e/test/long-running/multisig-wallet/multisig-delegation-rewards.test.ts:278]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/992/commits/8664427d551aebded9b0979446284bf483dc3b91

**Review by mkazlauskas — APPROVED** [review]
Great work! :rocket:

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Please bump the version in ChainHistory/openApi.json

**Comment by AngelCastilloB** [comment]
> Please bump the version in ChainHistory/openApi.json and regenerate versions.json by building cardano-services-client package.

done

**Review by mkazlauskas — APPROVED** [review]

**Pull request merged**